### PR TITLE
[FIX] website: unescape URL parameters when using them in autocomplete

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] when the search box autocomplete was introduced, the URL parameters are implicitly included into the RPC that fetches the autocompletion results.
Those parameters were not correctly unescaped before being sent to the RPC call.
Because of this, a timestamp such as "2023-01-01 23:00:00" was sent as "2023-01-01+23%3A00%3A00" to the server. If that string reached the SQL layer, the "+" was interpreted as defining a timezone.

This commit unescapes the URL parameters before using them in the RPC. Note that javascript's `decodeURIComponent` does not handle the '+' encoding of spaces inside URL parameters.

For testing purpose, the following updates were needed to make it possible to select the `<option>` within the Archive month `<select>`:
- because the `option`s are in a tree, the tool was adapted to take all `option`s into consideration instead of only the direct children of the `select`.
- because the `option` text is dynamically created from the date of the test execution, the tool was adapted to allow targeting an `option` based on its index by specifying the tour step's `run` as `'text index N'`, `N` being the index of the `option`.

Steps to reproduce:
- Enable the sidebar of the `/blog` page.
- Select a month in the sidebar.
- Type something in the search box.

=> Did show an error popup while obtaining the autocompletion records.

[1]: https://github.com/odoo/odoo/commit/7559626c54e34b41e1549e28276a650accec6986

task-3213916

### CODE REVIEW

**Comment by bso-odoo** [comment]
@rdeodoo Added a test checking autocomplete works when a timestamp is part of the URL parameters.

File: addons/website_blog/tests/test_ui.py [hunk 1]
```diff
@@ -23,3 +23,12 @@ def test_admin(self):
...
     def test_blog_post_tags(self):
         self.start_tour("/blog", 'blog_tags', login='admin')
+
+    def test_autocomplete_with_date(self):
+        self.env.ref('website_blog.opt_blog_sidebar_show').write({
+            'active': True,
+        })
+        self.env.ref('website_blog.opt_sidebar_blog_index_follow_us').write({
+            'active': False,
+        })
```
Comment (rdeodoo) [inline addons/website_blog/tests/test_ui.py:33]:
```suggestion
        self.env.ref('website_blog.opt_blog_sidebar_show').active = True
        self.env.ref('website_blog.opt_sidebar_blog_index_follow_us').active = False
```

File: addons/web_tour/static/src/js/running_tour_action_helper.js [hunk 2]
```diff
@@ -95,14 +95,19 @@ var RunningTourActionHelper = core.Class.extend({
...
         } else if (values.$element.is("select")) {
-            var $options = values.$element.children("option");
+            var $options = values.$element.find("option");
             $options.prop("selected", false).removeProp("selected");
             var $selectedOption = $options.filter(function () { return $(this).val() === text; });
             if ($selectedOption.length === 0) {
                 $selectedOption = $options.filter(function () { return $(this).text().trim() === text; });
             }
+            if ($selectedOption.length === 0 && text.startsWith('index ')) {
+                const index = parseInt(text.substring(6));
```
Comment (rdeodoo) [inline addons/web_tour/static/src/js/running_tour_action_helper.js:105]: Will need something more robust here, what if you want the 11th one?
Either a regex or assuming nothing else should be after `index ` and just take everything, maybe even split on `index `

File: addons/web_tour/static/src/js/running_tour_action_helper.js [hunk 3]
```diff
@@ -95,14 +95,19 @@ var RunningTourActionHelper = core.Class.extend({
...
+                $selectedOption = $options.slice(index, index + 1);
+            }
             $selectedOption.prop("selected", true);
             this._click(values);
+            values.$element.trigger("input");
```
Comment (rdeodoo) [inline addons/web_tour/static/src/js/running_tour_action_helper.js:110]: Why is this needed?

File: addons/website/static/src/snippets/s_searchbar/000.js [hunk 4]
```diff
@@ -65,15 +65,16 @@ publicWidget.registry.searchBar = publicWidget.Widget.extend({
             for (const keyValue of urlParams.split('&')) {
                 const [key, value] = keyValue.split('=');
                 if (value && key !== 'search') {
-                    this.options[key] = value;
+                    // Decode URI parameters: revert + to space then decodeURIComponent.
+                    this.options[decodeURIComponent(key.replace(/\+/g, ' '))] = decodeURIComponent(value.replace(/\+/g, ' '));
```
Comment (rdeodoo) [inline addons/website/static/src/snippets/s_searchbar/000.js:69]:
```suggestion
                    this.options[decodeURIComponent(key.replace(/\+/g, ''%20''))] = decodeURIComponent(value.replace(/\+/g, ''%20''));
```
No?

File: addons/website_blog/static/tests/tours/blog_search_with_date.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    extra_trigger: '#o_wblog_posts_loop span:has(i.fa-calendar-o):has(a[href="/blog"])',
+    run: 'text a',
+}, {
+    content: "Wait for suggestions",
+    trigger: '.o_searchbar_form .dropdown-menu .o_search_result_item',
+    run: () => {}, // This is a check.
+}, {
+    content: "Click on search icon",
+    trigger: '.o_searchbar_form button:has(i.fa-search)',
+}, {
```
Comment (rdeodoo) [inline addons/website_blog/static/tests/tours/blog_search_with_date.js:27]: you can merge both
```suggestion
    content: "Wait for suggestions",
    extra_trigger: '.o_searchbar_form .dropdown-menu .o_search_result_item',
    trigger: '.o_searchbar_form button:has(i.fa-search)',
}, {
```

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I didn't know that `text` could actually "click" on a select option, let alone that it would have fallback: first checking if `val()` match then if not, if `text` match.

I was thinking first that we could go with the same idea and make it explicit: `run: 'text 1'` so we add a third fallback: val > text > index with the same notation: you don't write `text val xx` or `text text xx`.

But it would conflict with an option having value or text 1 I guess, but seems to already be the case with options and value.
All in all, `index` is fine :+1: 

Note that in master I'd have probably made it more option like `select` or `option` instead of `text` but it's not really in our scope and I heard it was going to be refactored anyway (if not already with owl).

File: addons/web_tour/static/src/js/running_tour_action_helper.js (hunk 3 above)
Comment (bso-odoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js:110]: Because this is how the blog archive month `<select>`'s behavior is defined. https://github.com/odoo/odoo/blob/16.0/addons/website_blog/views/website_blog_components.xml#L60

File: addons/web_tour/static/src/js/running_tour_action_helper.js (hunk 2 above)
Comment (bso-odoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js]: Rewritten using a regex

**Comment by bso-odoo** [comment]
@rdeodoo Ready for review.
Applied most suggested changes. I do not think I can avoid the `input` event in stable.

File: addons/web_tour/static/src/js/running_tour_action_helper.js (hunk 3 above)
Comment (rdeodoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js:111]: Fair, I don't know if it's obvious to anyone when reading the code that this line is about triggering our custom owl `oninput`.
I let you judge if it's worth a comment.

Comment (bso-odoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js:112]: Added a comment.

**Comment by bso-odoo** [comment]
@rdeodoo Ready for review. Added a comment for the `trigger('input')`.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Was about to r+ but 2 remarks

File: addons/web_tour/static/src/js/running_tour_action_helper.js [hunk 6]
```diff
@@ -95,14 +95,21 @@ var RunningTourActionHelper = core.Class.extend({
...
+            var $options = values.$element.find("option");
             $options.prop("selected", false).removeProp("selected");
             var $selectedOption = $options.filter(function () { return $(this).val() === text; });
             if ($selectedOption.length === 0) {
                 $selectedOption = $options.filter(function () { return $(this).text().trim() === text; });
             }
+            const regex = /index\s+([0-9]+)/;
+            if ($selectedOption.length === 0 && regex.test(text)) {
+                const index = parseInt(regex.exec(text)[1]);
+                $selectedOption = $options.slice(index, index + 1);
```
[the remark this answers is not in the document]
Comment (bso-odoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js]: Adjusted syntax to `text option <N>` with `<N>` being 1-based, and extracted `$selectedOption` from its single index.

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/web_tour/static/src/js/running_tour_action_helper.js]: Alt Text

**Review by rdeodoo — APPROVED** [review]
@robodoo r+ ty benoit
